watcher/src/main/java/io/streamzi/openshift/FlowController.java

```diff
@@ -27,125 +30,149 @@ public FlowController() {
+            }
 
-                List<ConfigMap> existingCMs = osClient.configMaps().inNamespace(osClient.getNamespace()).withLabel("app", flow.getName()).list().getItems();
-                for (ConfigMap existing : existingCMs) {
+        } catch (Exception e) {
+            logger.log(Level.SEVERE, "Error parsing JSON flow data: " + e.getMessage(), e);
+        }
+    }
 
-                    if (!(existing.getMetadata().getLabels().containsKey("streamzi.io/kind") && existing.getMetadata().getLabels().get("streamzi.io/kind").equals("flow"))) {
-                        boolean found = false;
-                        for (ConfigMap newCM : deployer.getTopicMaps()) {
-                            if (existing.getMetadata().getName().equals(newCM.getMetadata().getName())) {
-                                found = true;
-                            }
-                        }
+    private void delete(Flow customResource) {
 
-                        if (!found) {
-                            logger.info("Deleting ConfigMap: " + existing.getMetadata().getName());
-                            osClient.configMaps().inNamespace(osClient.getNamespace()).withName(existing.getMetadata().getName()).delete();
-                        }
-                    }
-                }
+        try {
+            ProcessorFlow flow = new ProcessorFlow(customResource.getSpec());
+            logger.info("Flow Parsed OK");
 
-                //remove the flow CM if the flow is empty
-                if(flow.getNodes().size() == 0 && flow.getLinks().size() == 0){
-                    osClient.configMaps().inNamespace(osClient.getNamespace()).withName(flow.getName() + ".cm").delete();
-                }
+            //todo: is this safe and/or the best way to do it. Other apps *could* be deleted.
+            //remove DCs that are no longer required.
+            List<DeploymentConfig> existingDCs = osClient.deploymentConfigs().inNamespace(osClient.getNamespace()).withLabel("app", flow.getName()).list().getItems();
+
+            for (DeploymentConfig existingDC : existingDCs) {
+                osClient.deploymentConfigs().inNamespace(osClient.getNamespace()).withName(existingDC.getMetadata().getName()).delete();
+            }
 
-            } catch (Exception e) {
-                logger.log(Level.SEVERE, "Error parsing JSON flow data: " + e.getMessage(), e);
+            //todo: work out why the Strimzi topic CMs aren't being deleted - they get recreated even after the containers have gone away
+            List<ConfigMap> existingCMs = osClient.configMaps().inNamespace(osClient.getNamespace()).withLabel("app", flow.getName()).list().getItems();
+            for (ConfigMap existing : existingCMs) {
+                osClient.configMaps().inNamespace(osClient.getNamespace()).withName(existing.getMetadata().getName()).delete();
             }
```

Comment:
I think the stuff in the for-loop can be optimised using functional APIs from java8